File: georocket-server/src/main/java/io/georocket/index/xml/GmlIdIndexer.java
package io.georocket.index.xml;

import java.util.HashSet;
import java.util.Map;

import javax.xml.stream.events.XMLEvent;

import com.google.common.collect.ImmutableMap;

import io.georocket.util.XMLStreamEvent;

/**
 * Indexes GML IDs
 * @author Michel Kraemer
 */
public class GmlIdIndexer implements XMLIndexer {
  private static final String NS_GML = "http://www.opengis.net/gml";
  
  protected HashSet<String> ids = new HashSet<>();
  
  @Override
  public void onEvent(XMLStreamEvent event) {
    if (event.getEvent() == XMLEvent.START_ELEMENT) {
      String gmlId = event.getXMLReader().getAttributeValue(NS_GML, "id");
      if (gmlId != null) {
        ids.add(gmlId);
      }
    }
  }

  @Override
  public Map<String, Object> getResult() {
    return ImmutableMap.of("gmlIds", ids);
  }
}

File: georocket-server/src/main/kotlin/io/georocket/index/xml/GmlIdIndexer.kt
package io.georocket.index.xml

import io.georocket.util.XMLStreamEvent
import javax.xml.stream.events.XMLEvent

/**
 * Indexes GML IDs
 * @author Michel Kraemer
 */
class GmlIdIndexer : XMLIndexer {
  companion object {
    private const val NS_GML = "http://www.opengis.net/gml"
    private const val NS_GML_3_2 = "http://www.opengis.net/gml/3.2"
  }

  private var ids = mutableSetOf<String>()
  private var firstNsToCheck = NS_GML
  private var secondNsToCheck = NS_GML_3_2

  override fun onEvent(event: XMLStreamEvent) {
    if (event.event == XMLEvent.START_ELEMENT) {
      var gmlId = event.xmlReader.getAttributeValue(firstNsToCheck, "id")
      if (gmlId == null) {
        gmlId = event.xmlReader.getAttributeValue(secondNsToCheck, "id")
        if (gmlId != null) {
          // improve performance, by checking the second NS first
          val t = secondNsToCheck
          secondNsToCheck = firstNsToCheck
          firstNsToCheck = t
        }
      }
      if (gmlId != null) {
        ids.add(gmlId)
      }
    }
  }

  override fun getResult() = mapOf("gmlIds" to ids.toList())
}
